Добавить CD (#51)

birthdaysgift opened the pull request:
Мы хотим автоматизировать процесс доставки новых изменений на сервер. Для этого было принято решение настроить простой Continuous Delivery пайплайн.

В будущем мы планируем иметь два стенда - для тестирования (test) и для конечных пользователей (prod). И так как проект сейчас находится в стадии разработки, то для нас более актуален тестовый стенд, поэтому в рамках этой задачи мы будем настраивать деплой на тестовый стенд.

Далее будет общее описание изменений, которые нужно будет сделать в репозитории:

- Добавить uvicorn (если ещё не добавлен)
- Добавить gunicorn и прописать конфиг к нему
- Добавить `Dockerfile`, `docker-compose.yml` и `env.sh` для тестового стенда
- В папку рядом с докерфайлом добавить `Dockerfile.dockerignore` в котором исключить секреты и лишние папки кэша вспомогательных тулов (типо линтеров и пайтеста)
- Добавить nginx в докер компоуз и конфиг к нему
- Добавить гитхаб воркфлоу, которое будет разворачивать докер-компоуз на vps через ssh при пуше тега test
- убедиться, что приложение будет слушать 80-й порт, т.к. на 80-ом порте будет сервер, который раздаёт фронт

Далее приведены шаги, которые помогут настроить vps. Конечно, эти шаги могут отличаться в зависимости от конкретной vps, но их описание здесь может дать общее представление о том, что нужно сделать при разворачивании проекта на vps.

- Установить на VPS последнюю версию Docker Desktop по гайду из офф [доки](https://docs.docker.com/desktop/install/ubuntu/#install-docker-desktop)

  Чтобы скопировать установщик докера на vps нужно использовать что-то такое:
  ```bash
  scp ./docker-desktop-4.19.0-amd64.deb root@185.154.195.26:/downloads/docker-desktop-4.19.0-amd64.deb
  ```

- Проверить что docker работает
  ```bash
  docker container ls
  ```

  Если эта команда выше даёт ошибку связанную с сокетом докера, то [этот](https://stackoverflow.com/questions/43569781/unable-to-start-docker-service-with-error-failed-to-start-docker-service-unit/43576628#43576628) ответ на стэке может помочь. Там рекомендуют установить какой-то RHEL плагин такой командой:
  ```bash
  curl -sSL https://get.docker.com/ | sh
  ```

- Так как нам нужна возможность получать последние изменения проекта из гита на нашем vps сервере, то на туда необходимо установить гит (см офф [доку](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git))

Так как у гит-клиента установленного на vps должен быть доступ к нашему репозиторию, то необходимо настроить ssh ключи для соединения с гитхабом.

- Сгенерировать ssh ключ на VPS машине (см офф [доку](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent) гитхаба) и добавить его в свой (личный) гитхаб аккаунт.

- Склонировать репозиторий на VPS машину в папку /wlss-backend:
  ```
  git clone git@github.com:week-password/wisher-backend.git /wlss-backend
  ```

- перейти в папку с проектом `cd /wlss-backend` и настроить юзера для проекта (см офф [доку](https://git-scm.com/book/en/v2/Getting-Started-First-Time-Git-Setup#_your_identity) гита)

- Так как на vps машине будет полноценный гит репозиторий, мы хотим обезопасить себя от случайных коммитов с vps машины. Для этого необходимо запретить коммиты в репозиторий с помощью гит хука:
  ```bash
  echo "echo 'This repository is read-only'; exit 1;" > .git/hooks/pre-commit
  ```

Т.к. мы хотим, чтобы на стенд разворачивалась только та версия, которая нам нужна, а не любой коммит запушшеный в девелоп, было принято решение триггерить деплой на пуш определенного гит-тэга. Этот тэг будет виден всем разработчикам в гите, и они будут понимать какая именно версия и на какой стенд задеплоена в данный момент.

- С локальной машины создать тэг для тестового стенда и запушить его на ремоут:
  ```bash
  git push origin test --force
  ```

- На vps машине зафетчить этот тэг с ремоута и зачекаутится на него:
  ```bash
  git fetch origin +refs/tags/test:refs/tags/test
  git checkout test
  ```

- Перед тем как пытаться запустить наше приложение на vps, необходимо остановить apache сервер, который по дефолту может стоять на vps и слушать 80-й порт:
```bash
/etc/init.d/apache2 stop
```
  другие варианты этой команды можно посмотреть по [ссылке](https://www.cyberciti.biz/faq/star-stop-restart-apache2-webserver/)

- После этого необходимо убедится, что порт действительно освободился. Для этого надо из браузера зайти на айпи нашего сервера. Если при этом вы всё ещё видите стартовую страницу apache, то надо удалить `index.html` где-то внутри `/var/www`.

- Далее необходимо настроить окружение, в котором будет запускаться наше приложение. Для этого нужно сгенерировать `.env` файл
  ```bash
  source envs/test/env.sh
  ```
   и выставить корректные значения для секретов.

Процесс деплоя приложения будет управляться из воркфлоу гитхаба. Мы планируем в воркфлоу гитхаба вызывать команды типа `docker compose ... up/down` на нашем удалённом vps через ssh. Для этого нам понадобится [экшн](https://github.com/appleboy/ssh-action), который позволяет выполнять ssh команды из гитхаба.

- Чтобы этот экшн смог подсоединиться к нашему vps серверу по ssh, необходимо добавить в секреты репозитория на гитхабе креды для соединения с vps по ssh ([хост, юзер и пароль). Мы будем передавать эти секреты в ssh экшн через `${{ secrets.<SECRET_NAME> }}`

После этого можно пробовать запускать деплой с помощью пуша тэга `test`.

PS. Также мы решили пока не добавлять удаление ранов деплоя в cleanup-воркфлоу, потому что не совсем понятно, должны ли мы удалять раны для несуществующих коммитов, а если должны, то как. Ведь в случае удалении ветки, на которой стоит тэг, коммиты всё равно останутся под тэгом. А когда тэг будет переключен на другую ветку мы не сможем найти предыдущие раны с удалённой ветки, т.к. нам придётся проверить все раны тэга, а их со временем может накопиться очень много. К тому же нам придётся сверять эти раны с хэшами коммитов в репозитории, чтобы понять существует ли такой коммит. И в таком случае нам надо будет взять абсолютно все коммиты репозитория, ведь тэг не привязан ни к какой ветке и может быть в любой из них.

birthdaysgift commented on .github/workflows/lint.yml:
```diff
@@ -1,7 +1,10 @@
+on:
+  push:
+    branches:
+      - "*"
```
Исключаем пуш тэга для триггера линтера

birthdaysgift commented on .github/workflows/test.yml:
```diff
@@ -1,7 +1,10 @@
+on:
+  push:
+    branches:
+      - "*"
```
Иключаем пуш тэга для триггера тестов

birthdaysgift commented on envs/test/Dockerfile.dockerignore:
У этого файла необычное название. Так как он должен игнорить файлы в контексте конкретного докерфайла, который лежит в этой же папки.

Обычный `.dockerignore` в качестве контекста берёт ту папку, в которой находится. И так как у нас контекст сборки это рут директория приложения, то пришлось бы поместить его туда. Т.е. было бы:
```python
src/
  envs/
    test/
      Dockerfile
...
.dockerignore
```

Но так как у нас может быть больше одного окружения, которому потребуется докеригнор, то они будут конфликтовать, поэтому хотелось бы положить докеригнор в ту же папку, где находится докерфайл для конкретного окружения.

И сделать это можно, если назвать его `Dockerfile.dockerignore`. В таком случае он будет брать контекст сборки, в котором собирается `Dockerfile`, который лежит рядом с ним.

birthdaysgift commented:
После того как ПР будет влит, необходимо добавить в вики описание, как деплоить на стенд, а также шаги, необходимые для разворачивания на vps, описанные в этом тикете.

birthdaysgift commented:
@Rey-Lane проверить, что всё работает на стэнде можно тут: http://185.154.195.26/docs

Rey-Lane commented:
У меня 500 на эндпоинтах с entities

birthdaysgift commented:
> У меня 500 на эндпоинтах с entities

Судя по логам ты передал дефолтный бакет из сваггера - `string`. Но у нас в минио файлы для entity храняться в бакете `entity`. Поэтому он говорит, что бакета не существует.

![image](https://github.com/week-password/wlss-backend/assets/35929293/37421e1f-a20f-429c-ae06-d695cb67fbfb)


Я сделал POST запрос с телом `{"bucket": "entity", "key": "entity-1"}` и всё отработало без ошибок. После этого сделал `GET /entities/1` и тоже всё ок.

Rey-Lane commented:
Действительно, я прост тупой)

Rey-Lane reviewed (approved):
Выглядит, вроде бы, нормально. Трудно ревьюить такое, когда подобным не занимался. Доверяю твоему опыту и рабочему стенду 😅

birthdaysgift committed:
Добавить CD (#51)

Мы хотим автоматизировать процесс доставки новых изменений на сервер. Для этого было принято решение настроить простой Continuous Delivery пайплайн.

В будущем мы планируем иметь два стенда - для тестирования (test) и для конечных пользователей (prod). И так как проект сейчас находится в стадии разработки, то для нас более актуален тестовый стенд, поэтому в рамках этой задачи мы будем настраивать деплой на тестовый стенд.

Далее будет общее описание изменений, которые нужно будет сделать в репозитории:

- Добавить uvicorn (если ещё не добавлен)
- Добавить gunicorn и прописать конфиг к нему
- Добавить `Dockerfile`, `docker-compose.yml` и `env.sh` для тестового стенда
- В папку рядом с докерфайлом добавить `Dockerfile.dockerignore` в котором исключить секреты и лишние папки кэша вспомогательных тулов (типо линтеров и пайтеста)
- Добавить nginx в докер компоуз и конфиг к нему
- Добавить гитхаб воркфлоу, которое будет разворачивать докер-компоуз на vps через ssh при пуше тега test
- убедиться, что приложение будет слушать 80-й порт, т.к. на 80-ом порте будет сервер, который раздаёт фронт

Далее приведены шаги, которые помогут настроить vps. Конечно, эти шаги могут отличаться в зависимости от конкретной vps, но их описание здесь может дать общее представление о том, что нужно сделать при разворачивании проекта на vps.

- Установить на VPS последнюю версию Docker Desktop по гайду из офф [доки](https://docs.docker.com/desktop/install/ubuntu/#install-docker-desktop)

  Чтобы скопировать установщик докера на vps нужно использовать что-то такое:
  ```bash
  scp ./docker-desktop-4.19.0-amd64.deb root@185.154.195.26:/downloads/docker-desktop-4.19.0-amd64.deb
  ```

- Проверить что docker работает
  ```bash
  docker container ls
  ```

  Если эта команда выше даёт ошибку связанную с сокетом докера, то [этот](https://stackoverflow.com/questions/43569781/unable-to-start-docker-service-with-error-failed-to-start-docker-service-unit/43576628#43576628) ответ на стэке может помочь. Там рекомендуют установить какой-то RHEL плагин такой командой:
  ```bash
  curl -sSL https://get.docker.com/ | sh
  ```

- Так как нам нужна возможность получать последние изменения проекта из гита на нашем vps сервере, то на туда необходимо установить гит (см офф [доку](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git))

Так как у гит-клиента установленного на vps должен быть доступ к нашему репозиторию, то необходимо настроить ssh ключи для соединения с гитхабом.

- Сгенерировать ssh ключ на VPS машине (см офф [доку](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent) гитхаба) и добавить его в свой (личный) гитхаб аккаунт.

- Склонировать репозиторий на VPS машину в папку /wlss-backend:
  ```
  git clone git@github.com:week-password/wisher-backend.git /wlss-backend
  ```

- перейти в папку с проектом `cd /wlss-backend` и настроить юзера для проекта (см офф [доку](https://git-scm.com/book/en/v2/Getting-Started-First-Time-Git-Setup#_your_identity) гита)

- Так как на vps машине будет полноценный гит репозиторий, мы хотим обезопасить себя от случайных коммитов с vps машины. Для этого необходимо запретить коммиты в репозиторий с помощью гит хука:
  ```bash
  echo "echo 'This repository is read-only'; exit 1;" > .git/hooks/pre-commit
  ```

Т.к. мы хотим, чтобы на стенд разворачивалась только та версия, которая нам нужна, а не любой коммит запушшеный в девелоп, было принято решение триггерить деплой на пуш определенного гит-тэга. Этот тэг будет виден всем разработчикам в гите, и они будут понимать какая именно версия и на какой стенд задеплоена в данный момент.

- С локальной машины создать тэг для тестового стенда и запушить его на ремоут:
  ```bash
  git push origin test --force
  ```

- На vps машине зафетчить этот тэг с ремоута и зачекаутится на него:
  ```bash
  git fetch origin +refs/tags/test:refs/tags/test
  git checkout test
  ```

- Перед тем как пытаться запустить наше приложение на vps, необходимо остановить apache сервер, который по дефолту может стоять на vps и слушать 80-й порт:
```bash
/etc/init.d/apache2 stop
```
  другие варианты этой команды можно посмотреть по [ссылке](https://www.cyberciti.biz/faq/star-stop-restart-apache2-webserver/)

- После этого необходимо убедится, что порт действительно освободился. Для этого надо из браузера зайти на айпи нашего сервера. Если при этом вы всё ещё видите стартовую страницу apache, то надо удалить `index.html` где-то внутри `/var/www`.

- Далее необходимо настроить окружение, в котором будет запускаться наше приложение. Для этого нужно сгенерировать `.env` файл
  ```bash
  source envs/test/env.sh
  ```
   и выставить корректные значения для секретов.

Процесс деплоя приложения будет управляться из воркфлоу гитхаба. Мы планируем в воркфлоу гитхаба вызывать команды типа `docker compose ... up/down` на нашем удалённом vps через ssh. Для этого нам понадобится [экшн](https://github.com/appleboy/ssh-action), который позволяет выполнять ssh команды из гитхаба.

- Чтобы этот экшн смог подсоединиться к нашему vps серверу по ssh, необходимо добавить в секреты репозитория на гитхабе креды для соединения с vps по ssh ([хост, юзер и пароль). Мы будем передавать эти секреты в ssh экшн через `${{ secrets.<SECRET_NAME> }}`

После этого можно пробовать запускать деплой с помощью пуша тэга `test`.

PS. Также мы решили пока не добавлять удаление ранов деплоя в cleanup-воркфлоу, потому что не совсем понятно, должны ли мы удалять раны для несуществующих коммитов, а если должны, то как. Ведь в случае удалении ветки, на которой стоит тэг, коммиты всё равно останутся под тэгом. А когда тэг будет переключен на другую ветку мы не сможем найти предыдущие раны с удалённой ветки, т.к. нам придётся проверить все раны тэга, а их со временем может накопиться очень много. К тому же нам придётся сверять эти раны с хэшами коммитов в репозитории, чтобы понять существует ли такой коммит. И в таком случае нам надо будет взять абсолютно все коммиты репозитория, ведь тэг не привязан ни к какой ветке и может быть в любой из них.
